src/main/java/org/apache/tika/parser/external/WaitingExternalParser.java

```diff
@@ -15,48 +13,391 @@
+        try {
+            xhtml.startDocument();
+            xhtml.startElement("p");
+            char[] buffer = new char[1024];
+            for (int n = reader.read(buffer); n != -1; n = reader.read(buffer)) {
+                xhtml.characters(buffer, 0, n);
+            }
+            xhtml.endElement("p");
+            xhtml.endDocument();
+        } finally {
+            reader.close();
+        }
+    }
+
+    /**
+     * Starts a thread that sends the contents of the given input stream
+     * to the standard input stream of the given process. Potential
+     * exceptions are ignored, and the standard input stream is closed
+     * once fully processed. Note that the given input stream is <em>not</em>
+     * closed by this method.
+     *
+     * @param process process
+     * @param stream input stream
+     */
+    private void sendInput(final Process process, final InputStream stream) {
+        new Thread() {
+            public void run() {
+                OutputStream stdin = process.getOutputStream();
+                try {
+                    IOUtils.copy(stream, stdin);
+                } catch (IOException e) {
+                }
+            }
+        }.start();
+    }
+
+    /**
+     * Starts a thread that reads and discards the contents of the
+     * standard stream of the given process. Potential exceptions
+     * are ignored, and the stream is closed once fully processed.
+     *
+     * @param process process
+     */
+    private void ignoreStream(final InputStream stream) {
+        new Thread() {
+            public void run() {
+                try {
+                    IOUtils.copy(stream, new NullOutputStream());
+                } catch (IOException e) {
+                } finally {
+                    IOUtils.closeQuietly(stream);
+                }
+            }
+        }.start();
+    }
+    
+    private class Parser extends Thread
+    {
+        public volatile boolean isDone = false;
+        public volatile IOException ioException = null;
+    }
+    
+    private Parser extractMetadata(final InputStream stream, final Metadata metadata) {
+        Parser parser = new Parser() {
+            public void run() {
+               BufferedReader reader = new BufferedReader(new InputStreamReader(stream));
+               try {
+                  String line;
+                  while ( (line = reader.readLine()) != null ) {
+                     for(Pattern p : metadataPatterns.keySet()) {
+                        Matcher m = p.matcher(line);
+                        if(m.find()) {
+                           metadata.add( metadataPatterns.get(p), m.group(1) );
+                        }
+                     }
+                  }
+               } catch (IOException e) {
+                   this.ioException = e;
+               } finally {
+                  IOUtils.closeQuietly(reader);
+                  IOUtils.closeQuietly(stream);
+                  this.isDone = true;
```

Comment:
We could have gotten here after an IOException so this isn't a true measure of 'done'.